sdk/go/common/apitype/search.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apitype
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+type ResourceSearchResponse struct {
+	Total        *int64                    `json:"total,omitempty"`
+	Resources    []ResourceResult          `json:"resources,omitempty"`
+	Aggregations map[string]Aggregation    `json:"aggregations,omitempty"`
+	Pagination   *ResourceSearchPagination `json:"pagination,omitempty"`
+}
+
+func (r *ResourceSearchResponse) Error() error {
+	return fmt.Errorf("just a normal error")
+}
+
+// ResourceResult is the user-facing type for our indexed resources.
+//
+// If you add a property here, don't forget to update fieldMappings to make it
+// queryable!
+type ResourceResult struct {
+	Created      *string                      `json:"created,omitempty" csv:"created"`
+	Custom       *bool                        `json:"custom,omitempty" csv:"custom"`
+	Delete       *bool                        `json:"delete,omitempty" csv:"delete"`
+	Dependencies []string                     `json:"dependencies,omitempty" csv:"-"`
+	ID           *string                      `json:"id,omitempty" csv:"id"`
+	Matches      map[string][]json.RawMessage `json:"matches,omitempty" csv:"-"`
+	Modified     *string                      `json:"modified,omitempty" csv:"modified"`
+	Module       *string                      `json:"module" csv:"module"`
+	Name         *string                      `json:"name,omitempty" csv:"name"`
+	Package      *string                      `json:"package" csv:"package"`
+	Parent       *string                      `json:"parent.urn,omitempty" csv:"parent_urn"`
```

Comment:
This JSON tag won't work. It doesn't support descending into children like this.
You need a nested struct for the `Parent`, and a corresponding `URN`.

https://go.dev/play/p/t53DwbnG-2j

e.g.

```go
type ResourceResult struct {
  // ...
  Parent struct { URN string `json:"urn"` } `json:"parent"`
}
```

Or you can declare a `type StructURN string` that implements [json.Unmarshaler](https://pkg.go.dev/encoding/json#Unmarshaler) to extract the URN from a struct field.
Maybe something like this:

```go
type ParentURN string

func (p *ParentURN) UnmarshalJSON(bs []byte) error {
  var inner struct{ URN string `json:"urn"` }
  err := json.Unmarshal(bs, &inner)
  if err != nil { return err }
  *p = inner.URN
  return nil
}
```